pkg/resourceutils/scaleDown.go

```diff
@@ -0,0 +1,270 @@
+package resourceutils
+
+import (
+	"context"
+	"fmt"
+	"github.com/libopenstorage/stork/pkg/apis/stork/v1alpha1"
+	"github.com/portworx/sched-ops/k8s/core"
+	"strconv"
+	"strings"
+
+	migration "github.com/libopenstorage/stork/pkg/migration/controllers"
+	storkops "github.com/portworx/sched-ops/k8s/stork"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	k8sdynamic "k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+)
+
+func ScaleDownGivenResources(namespaces []string, resourceMap map[string]map[metav1.GroupVersionKind]map[string]string, config *rest.Config) error {
+	storkClient, err := storkops.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	crdList, err := storkClient.ListApplicationRegistrations()
+	if err != nil {
+		return err
+	}
+
+	for _, ns := range namespaces {
+		// 1. ApplicationResources
+		for _, gvk := range GetGVKsRelevantForScaling() {
+			resources, dynamicClient, err := ListResourcesByGVK(gvk, ns, config)
+			if err != nil {
+				return err
+			}
+			if resources == nil {
+				continue
+			}
+			for _, object := range resources.Items {
+				if _, present := resourceMap[ns][gvk][object.GetName()]; present {
+					err := scaleDownApplicationResource(dynamicClient, object, gvk)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		// 2. CronJobs
+		resources, dynamicClient, err := ListResourcesByGVK(cronJobGVK, ns, config)
+		if err != nil {
+			return err
+		}
+		if resources != nil {
+			for _, object := range resources.Items {
+				if _, present := resourceMap[ns][cronJobGVK][object.GetName()]; present {
+					err := scaleDownCronJob(dynamicClient, object)
+					if err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		// 3. CRD Resources
+		for _, applicationRegistration := range crdList.Items {
+			for _, crd := range applicationRegistration.Resources {
+				gvk := crd.GroupVersionKind
+				resources, dynamicClient, err := ListResourcesByGVK(gvk, ns, config)
+				if err != nil {
+					return err
+				}
+				if resources == nil {
+					continue
+				}
+				for _, object := range resources.Items {
+					if _, present := resourceMap[ns][gvk][object.GetName()]; present {
+						err := scaleDownCRDResource(dynamicClient, object, crd, config)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func scaleDownApplicationResource(dynamicClient k8sdynamic.ResourceInterface, object unstructured.Unstructured, resourceType metav1.GroupVersionKind) error {
+	content := object.UnstructuredContent()
+	currentReplicas, found, err := unstructured.NestedInt64(content, "spec", "replicas")
+	if err != nil {
+		return err
+	}
+	if !found {
+		currentReplicas = 1
+	}
```

Comment:
Explain this.

Reply:
I have followed the scaling down logic from prepareApplicationResources method in the migration flow.
https://github.com/libopenstorage/stork/blob/5062314469757f8e10b41a6c97c693c07995069c/pkg/migration/controllers/migration.go#L1474